Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+        }
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.serializeRecord(record) ?? "").utf8.count
+        }
+
+        let sizeOfRecords = records.reduce(0, combine: sizeReducer)
+
+        // Batch up all of the records and fail early if there is an issue
+        let batchingResult = batchesFromRecords(records)
+        guard let batches = batchingResult.successValue else {
+            let failure = batchingResult.failureValue!
+            log.debug("Unable to generate batches from records submitted to batch client: \(failure)")
+            return deferMaybe(failure)
+        }
+
+        // We have too many records for a single post so we'll either need to upload in a single batch or
+        // multiple batches
+        if records.count > config.maxTotalRecords || sizeOfRecords > config.maxTotalBytes {
+            return self.bunchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        if records.count > config.maxPostRecords || sizeOfRecords > config.maxPostBytes {
+            return self.batchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        // We can just do a single post instead of batching
+        log.debug("Batch fits within a single request. Submitting records in a single post.")
+        let lines = optFilter(records.map(self.serializeRecord))
+        return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: nil)
+            >>== effect({ onCollectionUploaded($0.value) })
+            >>> succeed
+    }
+
+    public func bunchUpload(batch: [[String]], ifUnmodifiedSince: Timestamp?, onCollectionUploaded: (POSTResult -> Void)) -> Success {
+
+        // Need to break apart the provided batch into multiple batches (AKA a bunch) so we can fit them in batch calls
+        var mutBatch = batch
+
+        /* 
+         * When talking about a 'batch' and multiple 'batches' the word becomes very confusing. In this context,
+         *
+         * A 'batch' consists of N requests to the server beginning with a batch=true call and ends with a commit=true call.
+         * A 'bunch' is many of these calls.
+         * A partialBatch is one of the N requests made within a 'batch'.
+         *
+         * Here are some typealiases to make your life easier.
+         */
+
+        typealias Batch = [[String]]
+        typealias Bunch = [Batch]
+
+        var bunch = Bunch()
+
+        while !mutBatch.isEmpty {
+            var recordCount = 0
+            var size = 0
+            var smallerBatch = Batch()
+
+            while size <= config.maxTotalBytes && recordCount <= config.maxTotalRecords && !mutBatch.isEmpty {
+                let partialBatch = mutBatch.removeFirst()
```

Comment:
Thanks for the explanation! I completely disregarded the behavior of not calling commit and it's effect of not finishing the batch. This makes a lot of sense.

> Perhaps the best way to think about this is in terms of Clojure-esque channels and transducers. A channel is a producer-driven queue, like a conveyor belt. A transducer is a worker between two conveyor belts. In this case, our caller is pushing records onto the conveyor belt. We have a transducer that's bundling up records into collections of 50/1MB, and one that's shifting a conveyor belt into different bunches as a limit is reached (similar process). The consumer is uploading each bundle once it arrives. Finally, you tell the destination to commit when the channel is closed.

Great analogy by the way 👌